### PR TITLE
Create Submission Types Administration

### PR DESCRIPTION
**Pull request by MTsatsarov**
Closes https://github.com/SoftUni-Internal/exam-systems-issues/issues/1186

**Summary of the changes made**:
1. Created administration pages, validations, permissions and services.
2.

### CODE REVIEW

File: Services/Administration/OJS.Services.Administration.Business/SubmissionTypes/SubmissionTypesBusinessService.cs [hunk 1]
```diff
@@ -17,4 +18,43 @@ public SubmissionTypesBusinessService(ISubmissionTypesDataService submissionType
 
     public async Task<List<SubmissionTypesInProblemView>> GetForProblem() =>
         await this.submissionTypesDataService.GetAll().MapCollection<SubmissionTypesInProblemView>().ToListAsync();
+
+    public override Task<SubmissionTypeAdministrationModel> Get(int id) =>
+        this.submissionTypesDataService.GetByIdQuery(id).FirstOrDefaultAsync()
+            .Map<SubmissionTypeAdministrationModel>();
```
Comment (gogo4ds) [inline Services/Administration/OJS.Services.Administration.Business/SubmissionTypes/SubmissionTypesBusinessService.cs:24]: Use `MapCollection`

**Review by gogo4ds — CHANGES_REQUESTED** [review]

File: Services/Administration/OJS.Services.Administration.Business/SubmissionTypes/Validators/SubmissionTypeAdministrationModelValidator.cs [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+    {
+        var isValid = Enum.TryParse(typeof(T), value, out _);
+
+        if (isValid)
+        {
+            return true;
+        }
+
+        return false;
+    }
```
[the remark this answers is not in the document]
Comment (MTsatsarov) [reply Services/Administration/OJS.Services.Administration.Business/SubmissionTypes/Validators/SubmissionTypeAdministrationModelValidator.cs]: Done

File: Services/Administration/OJS.Services.Administration.Business/SubmissionTypes/SubmissionTypesBusinessService.cs (hunk 1 above)
Comment (MTsatsarov) [reply Services/Administration/OJS.Services.Administration.Business/SubmissionTypes/SubmissionTypesBusinessService.cs]: Done

File: Common/OJS.Common/Extensions/RuleBuilderExtensions.cs [hunk 3]
```diff
@@ -0,0 +1,24 @@
+﻿namespace OJS.Common.Extensions;
+
+using FluentValidation;
+using System;
+
+public static class RuleBuilderExtensions
+{
+    public static IRuleBuilderOptions<T, string?> MustBeValidEnum<T, TEnum>(this IRuleBuilder<T, string?> ruleBuilder)
+        where TEnum : Enum
+        => ruleBuilder.Must(BeValidEnum<TEnum>).WithMessage($"The value is not of type {typeof(TEnum)}");
```
Comment (gogo4ds) [inline Common/OJS.Common/Extensions/RuleBuilderExtensions.cs:10]: I think the exception message will not be very descriptive.  what will `typeof(TEnum)` be in string representation?

File: Services/Administration/OJS.Services.Administration.Business/SubmissionTypes/Validators/SubmissionTypeAdministrationModelValidator.cs [hunk 4]
```diff
@@ -0,0 +1,31 @@
...
+            .WithMessage("Id cannot be less than 0");
+
+        this.RuleFor(model => model.Name)
+            .NotNull()
+            .NotEmpty()
+            .WithMessage("Name field is required");
+
+        this.RuleFor(model => model.CompilerType)
+            .MustBeValidEnum<SubmissionTypeAdministrationModel, CompilerType>()
+            .WithMessage("Compiler type is invalid");
```
Comment (gogo4ds) [inline Services/Administration/OJS.Services.Administration.Business/SubmissionTypes/Validators/SubmissionTypeAdministrationModelValidator.cs:25]: We should not need the custom message, if we have good enough message in the default implementation..

**Review by gogo4ds — CHANGES_REQUESTED** [review]

File: Services/Administration/OJS.Services.Administration.Business/SubmissionTypes/SubmissionTypesBusinessService.cs [hunk 5]
```diff
@@ -17,4 +18,42 @@ public SubmissionTypesBusinessService(ISubmissionTypesDataService submissionType
 
     public async Task<List<SubmissionTypesInProblemView>> GetForProblem() =>
         await this.submissionTypesDataService.GetAll().MapCollection<SubmissionTypesInProblemView>().ToListAsync();
+
+    public override async Task<SubmissionTypeAdministrationModel> Get(int id) =>
+         (await this.submissionTypesDataService.GetByIdQuery(id).MapCollection<SubmissionTypeAdministrationModel>().FirstOrDefaultAsync())!;
```
Comment (gogo4ds) [inline Services/Administration/OJS.Services.Administration.Business/SubmissionTypes/SubmissionTypesBusinessService.cs:23]: Isn't this line too long?

File: Services/Administration/OJS.Services.Administration.Business/SubmissionTypes/Validators/SubmissionTypeAdministrationModelValidator.cs (hunk 4 above)
Comment (MTsatsarov) [reply Services/Administration/OJS.Services.Administration.Business/SubmissionTypes/Validators/SubmissionTypeAdministrationModelValidator.cs]: Fixed.

File: Common/OJS.Common/Extensions/RuleBuilderExtensions.cs (hunk 3 above)
Comment (MTsatsarov) [reply Common/OJS.Common/Extensions/RuleBuilderExtensions.cs]: Fixed.

File: Services/Administration/OJS.Services.Administration.Business/SubmissionTypes/SubmissionTypesBusinessService.cs (hunk 5 above)
Comment (MTsatsarov) [reply Services/Administration/OJS.Services.Administration.Business/SubmissionTypes/SubmissionTypesBusinessService.cs]: Fixed.

**Review by gogo4ds — APPROVED** [review]

**Pull request merged**